TAN-5196 Files in sensemaking

kogre opened the pull request.

Commit:
Split off LLM file preparation from Description generator service

Commit:
Support file attachments for Analysis::Analysis

Commit:
cover analysis attachment in spec

adessy committed:
[TAN-5196] Rename +LLMPreparation+ to +LLMFilePreprocessor+

adessy committed:
Merge branch 'master' into TAN-5196-files-in-sensemaking

# Conflicts:
#	back/app/models/files/file_attachment.rb

adessy committed:
[TAN-5196] Include files in Q&A analyses

adessy committed:
[TAN-5196] Add files relationship to serialized analyses

adessy committed:
[TAN-5196] Include files in AI analyses only if the feature flag is active

adessy committed:
[TAN-5196] Fix circular dependency when loading QAndAMethod classes

Loading `Analysis::QAndAMethod::OnePassLLM` before `Analysis::QAndAMethod::Base` was causing:
> NameError:
>    uninitialized constant Analysis::QAndAMethod::OnePassLLM

adessy committed:
[TAN-5196] Fix circular dependency when loading SummarizationMethod classes

Loading `Analysis::SummarizationMethod::OnePassLLM` before `Analysis::SummarizationMethod::Base` was causing:
> NameError:
>    uninitialized constant Analysis::SummarizationMethod::OnePassLLM

adessy committed:
[TAN-5196] Support file attachment for Analysis::Question

adessy committed:
[TAN-5196] Add the file_ids parameter to the `POST .../question` endpoint

adessy committed:
[TAN-5196] Use file IDs from analysis data when asking questions

adessy committed:
[TAN-5196] Add the files parameter to the PATCH `.../analysis/:id` endpoint

This parameter allows attaching multiple files to an analysis in a single request.

adessy committed:
[TAN-5196] Use new Q&A prompt when `data_repository_ai_analysis` is enabled

This will allow us to test it before releasing it more broadly.

adessy committed:
[TAN-5196] Correct value extraction for multiloc fields in XLSX exports

Ignore blank values (empty strings or whitespace) when extracting translations from multiloc fields. This also fixes input formatting issues in Sensemaking, which uses `Export::Xlsx::ValueVisitor`.

adessy committed:
[TAN-5196] Add support for image inputs to Azure OpenAI client

adessy committed:
[TAN-5196] Sort file options alphabetically in FileSelectionView component

adessy committed:
[TAN-5196] Use PDF preview as fallback for non-PDF files

Since Azure OpenAI API only accepts PDFs (and some images for vision models), we now fall back to the PDF preview (if available) when a non-PDF file is provided.

adessy committed:
[TAN-5196] Raise `UnsupportedAttachmentError` for unsupported image types in `AzureOpenAI`

adessy committed:
[TAN-5196] Fix typo

adessy committed:
[TAN-5196] Refine Q&A errors and defer retries to the job scheduler

By deferring the retry logic to the job scheduler, we can avoid suspending the thread.

adessy committed:
[TAN-5196] Add the "New" label to file attachments in AI Analysis

adessy committed:
[TAN-5196] Uncomment FE code related to file attachments in AI analysis

adessy committed:
[TAN-5196] Add `last_error_class` column to `analysis_background_tasks` table

adessy committed:
[TAN-5196] Add failure reason to serialized background tasks

adessy committed:
[TAN-5196] Persist the error class when background tasks fail

adessy committed:
[TAN-5196] Add `failure_reason` field to background task types

adessy committed:
[TAN-5196] Show error messages for background task failures

Display an error message when a background task fails in Sensemaking / Analysis AI.

adessy committed:
[TAN-5196] Tweak spacing in Insight component for consistency

adessy committed:
[TAN-5196] Handle "too many images" error in AI analysis

adessy committed:
[TAN-5196] Handle unsupported file error in AI analysis

adessy committed:
[TAN-5196] Refactor AI disclaimer in AI Analysis

Disclaimers used to show up for every question and summary. Now there's just one clear disclaimer at the top to keep things cleaner and less repetitive.

adessy committed:
[TAN-5196] Improve AI Disclaimer banner

Add a detailed explanation to the AI Disclaimer banner and make it collapsible.

adessy committed:
[TAN-5196] Disable auto-scrolling to errors in AI Analysis

adessy committed:
[TAN-5196] Remove unnecessary wrapper Box in Insights component

adessy committed:
[TAN-5196] Align spacing and layout of Summary with Question

adessy committed:
[TAN-5196] Rename showTooltip to showAiWarning

adessy committed:
[TAN-5196] Move insight icon before text when warning hidden

Make the icon (stars or question-bubble) appear before the text in the insight header when `showAiWarning` is false to make things look nicer.

adessy committed:
[TAN-5196] Refactor divider placement in Analysis AI

Moved dividers from individual insight components to the parent
container for a cleaner layout. Added a flat divider between the
controls and the insights list to better highlight the separation.

adessy committed:
[TAN-5196] Improve AI Disclaimer banner Adrien Dessy Today 17:22

adessy committed:
[TAN-5196] Tweak spacing and alignment in Insights components

adessy committed:
[TAN-5196] Revamp Insight footer

Aligned the footer icons with the file-upload icon size to make them appear more balanced with the Insight text. Also moved the icons to the right for a cleaner, more balanced visual.

adessy committed:
[TAN-5196] Update file management button in Insights

Clarify that the project files button opens in a new tab.

adessy committed:
[TAN-5196] Preserve analysis file attachments on reopen

Fix an issue where opening "Attach files" cleared the file list and triggered unnecessary PATCH requests. This ensures the current analysis files are retained and auto-submit only runs when the user makes actual changes.

adessy committed:
[TAN-5196] Fix @typescript-eslint/no-unnecessary-condition issue

adessy committed:
[TAN-5196] Add missing TooManyImagesError in Analysis::QAndAJob#handle_error

adessy committed:
[TAN-5196] Adjust gap spacing in Insights headers

Give the icon a bit more breathing room.

adessy commented:
Hey @amanda-anderson, could you please review the front-end part? 🥺 @kogre will check the back-end.

kogre commented on back/engines/commercial/analysis/app/models/analysis/question.rb:
```diff
@@ -26,12 +26,18 @@
     validates :accuracy, numericality: { in: 0..1 }, allow_blank: true
+
+    def source_project
+      insight.analysis.project
```
Shouldn't this be `insight.analysis.source_project`? For survey analyses, analysis.project is `nil`.

adessy replied:
> For survey analyses, analysis.project is nil.

How come? :)

adessy replied:
Also, and this is probably related, I wasn't sure why you added `source_project` or how it conceptually differs from `project`.

kogre commented on back/engines/commercial/analysis/config/prompts/q_and_a.v2.erb:
Curious about the prompt. Expecially wondering about the large emphasis on quantitative analysis, as it's something the model often struggles with (these plan calls can't use tools or code to respond, as opposed to using e.g. ChatGPT), there's a chance it might want to overreach?

Were you able to test this somewhat?

adessy replied:
I was planning to test it on the demo platforms once it's deployed. And yes, maybe there's a bit too much emphasis on "quantitative analysis", although to be fair, that was already quite present in the previous version:
> Your only goal is to answer the following question about these responses, as accurately and **as quantified as possible**

adessy replied:
But I'll add a new temporary setting to the feature flag so we can override the prompt. That should make testing easier.

kogre reviewed (commented):
I can't review my own PR either, so adding some thoughts as comments.

I only skimmed through it, but I think it makes sense. Next to my comments, also wondered whether we really need to integrate the attaching of files to an analysis with the analysis update endpoint, instead of using the attachment endpoints as we do in the other places in the app. I think it might be cleaner, but maybe I'm missing something.

I didn't check the FE code

adessy committed:
[TAN-5196] Fix typo

amanda-anderson commented on front/app/api/analysis_background_tasks/types.ts:
```diff
@@ -22,6 +22,7 @@ type SharedAttributes = {
   started_at: string | null;
   ended_at: string | null;
   state: 'queued' | 'in_progress' | 'succeeded' | 'failed';
+  failure_reason: 'unsupported_file_type' | 'too_many_images' | null;
```
@adessy Do you think we could include the file name in this error message? If there are multiple files attached it's a bit unclear maybe? Not sure on the effort though.

<img width="538" height="219" alt="image" src="https://github.com/user-attachments/assets/02164d6e-dbb3-49dd-a24b-b2e7ac8576fb" />

amanda-anderson replied:
Also if this could be raised when the user is first selecting the files, that would be ideal I think :) Not blocking though.

adessy replied:
@amanda-anderson Yes, that would be nice, but I don't think I'll get there this time. The inlined warning is already a good improvement that wasn't originally planned. As immediately warning the user when selecting files, it's a bit tricky to do it properly without bypassing the planning mechanism, which is meant to dynamically choose the best model for the task (and the models may have different restrictions).

I suggest we pause here for now and see if the feature gains some adoption. I don't think these small problems will stop users from using it if they find it valuable.

amanda-anderson commented on front/app/containers/Admin/projects/project/analysis/Insights/Files/FileSelectionView/index.tsx:
```diff
@@ -1,16 +1,18 @@
-import React, { useEffect } from 'react';
+import React, { useEffect, useRef } from 'react';
```
I have a couple suggestions for this file:
- Using lodash `isEqual`
- Using the files from the analysis `(analysis?.data.relationships.files?.data.map((file) => file.id)` directly when instantiating the `previousFileIdsRef` and `methods` so we can remove the useEffect (unless I've made a mistake and we do still need it).

I'll send the final code in a DM on Slack - might be good to test once more just to make sure I didn't miss anything though! :)

adessy replied:
Thanks! Your code was much better 🤩

amanda-anderson reviewed (approved):
A couple small comments, but happy to approve in advance! :rocket:

adessy committed:
[TAN-5196] Simplify FileSelectionView

Co-authored-by: Amanda <33987955+amanda-anderson@users.noreply.github.com>

adessy committed:
[TAN-5196] Add specs for Analysis::QAndAJob#handle_error

adessy committed:
[TAN-5196] Add specs for LLMFilePreprocessor

adessy committed:
[TAN-5196] Add custom prompt support for AI analysis

This allows admins to customize the AI analysis prompt via AdminHQ without requiring code changes.

adessy committed:
[TAN-5196] Add file support tests for AzureOpenAI

adessy commented:
@amanda-anderson This PR adds some copy. I added the branch to Crowdin, but it doesn’t seem to be picking up the strings. Do you have any idea what might be going wrong?

amanda-anderson commented:
> This PR adds some copy. I added the branch to Crowdin, but it doesn’t seem to be picking up the strings.

@adessy Not sure, but I've tried just re-triggering the build, and then I'll try to sync the branch from CrowdIn.

If you merge in master, maybe that could help too? Haven't encountered this issue myself tbh :thinking: 


EDIT: Yep that seems to have fixed it! Picky CrowdIn :sweat_smile:

Commit:
Translations updated by CI (extract-intl)

adessy commented:
@amanda-anderson thanks! 🫶

adessy committed:
Merge branch 'master' into TAN-5196-files-in-sensemaking

# Conflicts:
#	back/db/structure.sql

adessy committed:
[TAN-5196] Fix rubocop offenses
